### PR TITLE
[2572] Use dynamic handles

### LINKED ISSUE #2572: Use one handle per edges
Right now, we are defining 4 handles per nodes.
Each edge's path start or end on an handle.
When we have several edges that share the same handles, the layout is messy.

We need to have one handle per source/target of an edge.

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2572

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/ConvertEngine.types.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/ConvertEngine.types.ts]: Nitpick, but `gqlEdges` could be defined above, close to the existing `gqlNode`.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: Shouldn't these be chained, to avoid calling `onNodesChange` twice, the first time on a version of `changes` which will be made obsolete by `onHandleChange`?
Also, done this way `onHandleChange` sees the original version of `changes` without the changes performed by `onBorderChange`.

```js
onNodesChanges(onHandleChange(onBorderChange(changes)));
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useRefreshConnectionHandles.tsx
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useRefreshConnectionHandles.tsx]: Not sure this makes sense. A function that when invoked returns an object which... references the function itself?

`useRefreshConnectionHandles` is a plain JS function whose return value is never used. It's always invoked directly:
```
useRefreshConnectionHandles(id, data.connectionHandles);
```

`UseRefreshConnectionHandlesValue` could be removed, and the function simply be typed to return `void`.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useRefreshConnectionHandles.types.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useRefreshConnectionHandles.types.ts:1]: Not needed IMO.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts:33]: Shouldn't this be more like `||` instead of `&&`? Written this way, it seems like we will update the connection handles only if *both* the source and target positions change.

Maybe something like this.

```js
      if (
        nodeSourceConnectionHandle?.position !== sourcePosition ||
        nodeTargetConnectionHandle?.position !== targetPosition
      ) {
        let nodeSourceConnectionHandles: ConnectionHandle[] = sourceNode.data.connectionHandles;
        if (nodeSourceConnectionHandle?.position !== sourcePosition) {
          nodeSourceConnectionHandles = sourceNode.data.connectionHandles.map(
            (nodeConnectionHandle: ConnectionHandle) => {
              if (nodeConnectionHandle.id === sourceHandle && nodeConnectionHandle.type === 'source') {
                nodeConnectionHandle.position = sourcePosition;
              }
              return nodeConnectionHandle;
            }
          );
        }

        let nodeTargetConnectionHandles: ConnectionHandle[] = targetNode.data.connectionHandles;
        if (nodeTargetConnectionHandle?.position !== targetPosition) {
          nodeTargetConnectionHandles = targetNode.data.connectionHandles.map(
            (nodeConnectionHandle: ConnectionHandle) => {
              if (nodeConnectionHandle.id === targetHandle && nodeConnectionHandle.type === 'target') {
                nodeConnectionHandle.position = targetPosition;
              }
              return nodeConnectionHandle;
            }
          );
        }

...
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts]: Seems OK, but possibly slower than needed. We iterate on *all nodes* on the diagram (most of which are not concerned at all by the current edge) for every edge where the source/target position has changed.

It's probably OK for now to get something that works and is readable/obvious, but we might want to return to this if we see performance issues (especially if the whole layout, which invokes this, is called a lot more often to hande the resize issues).

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionHandles.tsx
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionHandles.tsx]: `key` => `position` would be more readable IMO.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertDiagram.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertDiagram.ts]: `usedHandles.find((usedHandles)` is confusing. The parameter in the lamba should not use the same name as the array we are iterating on.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertDiagram.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertDiagram.ts]: Same remark as above.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useHandleChange.tsx
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useHandleChange.tsx]: This looks very similar to what's in `layoutHandles` (except it uses `setNodes` instead of `diagram.nodes =`), so:
* it's probably possible to share most/all of the code;
* the same remark I did regarding the logic of the `if` applies here too.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertHandles.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertHandles.ts]: Strange to return strings from functions named `numberSomething`.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertHandles.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/converter/convertHandles.ts]: Maybe I'm missing something, but any reason why two plain counters wouldn't work?

```js
let sourceHandlesCounter = 0;
...
sourceHandlesCounter += 1;
...
id: `handle--source--${gqlNode.id}--${sourceHandlesCounter}`,
```

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts:33]: Yes currently, how we calculate a change in handle position is making both handles to change position.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/layoutHandles.ts]: Yes, this could be improved but we only call this is the handles position changed.
`      if (
        nodeSourceConnectionHandle?.position !== sourcePosition &&
        nodeTargetConnectionHandle?.position !== targetPosition
      ) {
`

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useHandleChange.tsx
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useHandleChange.tsx]: I pushed a version that reuse more code but since we are in 2 différents situations I don't see how I could share more code.

**Review by pcdavid — APPROVED** [review]
This looks fine to me. There are a few minor known issues left, but they can probably be handled afterwards. @sbegaudeau can have a look and see if you find anything worth blocking this? Otherwise I'll merge it and start working the last remaining task (the feedback).

Known issues:
1. the edges go 1/2 pixels too far inside the nodes instead of stopping right at their border.
2. the presence of the "virtual handles" (one on the left side, one one the right side), even if they are invisible, means a single edge on the left or right side of a node does not point to the middle of the node's side.
3. the way "sibling handles" are ordered on a given side does not take the position of the target nodes into account, which creates edge crossings in some cases. There's probably a way to improve this in common cases without going all the way and try to fix all potential edge crossings.
4. Feedback is missing (incompatible nodes faded, compatible nodes unfaded, selected/target node highlighed).
5. It's possible to trigger a reconnect even on an edge which is not currently selected (maybe we actually want this?).

There a probably also possible ergonomics improvements that we'll discover once this is merged and we actually start using all this.

**Comment by pcdavid** [comment]
Waiting for the build to merge. We'll handle the known issues and missing parts in later PRs.

**Pull request merged**
